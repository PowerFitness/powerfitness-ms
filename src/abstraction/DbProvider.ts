import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
