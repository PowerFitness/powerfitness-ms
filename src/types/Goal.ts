export interface Goal {
    id: number;
    planId: number;
    type: string;
    name: string;
    unit: string;
    value: number;
    createdDate: string;
    updatedDate: string;
}
