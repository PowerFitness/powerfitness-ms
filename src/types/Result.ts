export interface Result {
	id: number;
	userUniqueId: string;
	date: string;
	type: string;
	subtype: string;
	name: string;
	unit: string;
	value: number;
	createdDate: string;
	lastUpdatedDate: string;
}
