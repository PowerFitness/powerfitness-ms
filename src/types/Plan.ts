import { Goal } from './Goal';

export interface Plan {
	id: number;
	userUniqueId: string;
	motivationStatement: string;
	createdDate: string;
	lastUpdatedDate: string;
    goals?: Array<Goal>
}
