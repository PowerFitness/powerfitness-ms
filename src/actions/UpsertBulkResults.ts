import { Action } from '../abstraction/Action';
import DbProvider from '../abstraction/DbProvider';
import { OkPacket } from 'mysql';
import { Result } from '../types/Result';
import { buildBulkInsertValues } from '../util/buildConditions';

export class UpsertBulkResults extends Action<OkPacket> {
	results: Result[];
	constructor(dbProvider: DbProvider, results: Result[]) {
		super(dbProvider);
		this.results = results;
	}
	buildQuery(): string {
		return ` 
		insert into result
			(id, userUniqueId, date, type, subtype, name, unit, value, createdDate, lastUpdatedDate)
		values
		${buildBulkInsertValues([ 'id', 'userUniqueId', 'date', 'type', 'subtype', 'name', 'unit', 'value', 'createdDate', 'lastUpdatedDate' ],
		this.results)}			
		on duplicate key update
			value= VALUES(value), lastUpdatedDate = VALUES(lastUpdatedDate)`
	}
}

export default UpsertBulkResults;

