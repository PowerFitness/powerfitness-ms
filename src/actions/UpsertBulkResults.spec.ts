import DbProvider from '../abstraction/DbProvider';
import { Result } from '../types/Result';
import { mock } from 'jest-mock-extended';
import UpsertBulkResults from './UpsertBulkResults';

describe('UpsertBulkResults', () => {
	test('buildQuery', () => {
		const result: Array<Result> = [
			{
				id: 1,
				userUniqueId: 'uid',
				date: 'today',
				type: 'exercise',
				subtype: 'exercise',
				name: 'running',
				value: 30,
				unit: 'minutes',
				createdDate: 'today',
				lastUpdatedDate: 'today',
			},
			{
				id: 2,
				userUniqueId: 'uid',
				date: 'today',
				type: 'water',
				subtype: 'water',
				name: 'water',
				value: 80,
				unit: 'ounces',
				createdDate: 'today',
				lastUpdatedDate: 'today',
			},
			{
				id: 3,
				userUniqueId: 'uid',
				date: 'today',
				type: 'nutrition',
				subtype: 'dinner',
				name: 'pizza',
				value: 200,
				unit: 'calories',
				createdDate: 'today',
				lastUpdatedDate: 'today',
			}
		];
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: UpsertBulkResults = new UpsertBulkResults(dbProvider, result);
		expect(action.buildQuery()).toMatchSnapshot();
	})
})
