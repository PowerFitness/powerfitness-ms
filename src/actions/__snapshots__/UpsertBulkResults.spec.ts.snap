// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpsertBulkResults buildQuery 1`] = `
" 
		insert into result
			(id, userUniqueId, date, type, subtype, name, unit, value, createdDate, lastUpdatedDate )
		values
		(1, 'uid', 'today', 'exercise', 'exercise', 'running', 'minutes', 30, CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP()), (2, 'uid', 'today', 'water', 'water', 'water', 'ounces', 80, CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP()), (3, 'uid', 'today', 'nutrition', 'dinner', 'pizza', 'calories', 200, CURRENT_TIMESTAMP(), CURRENT_TIMESTAMP())			
		on duplicate key update
			value = VALUES(value), lastUpdatedDate=VALUES(lastUpdatedDate)"
`;
